### PR TITLE
Remove image via storage if a buildah container is associated

### PR DESCRIPTION
**Pull request by TomSweeneyRedHat**
Signed-off-by: TomSweeneyRedHat <tsweeney@redhat.com>

When a removal of an image returns ErrImageUsedByContainer, rather than raising an error immediately we get the image from the containers/storage and see if we can remove containers from there for that image (generally those containers are put into play by Buildah) and then try again to remove the image from Podman.

### CODE REVIEW

File: cmd/podman/rmi.go [hunk 1]
```diff
@@ -77,3 +98,217 @@ func rmiCmd(c *cli.Context) error {
 	}
 	return lastError
 }
+
+func rmiBuildahCmd(c *cli.Context, imageID string) error {
+
+	store, err := getStore(c)
```
Comment (mheon) [inline cmd/podman/rmi.go]: We already have a storage.Store in libpod, don't need this

File: cmd/podman/rmi.go [hunk 2]
```diff
@@ -77,3 +98,217 @@ func rmiCmd(c *cli.Context) error {
...
+	if err != nil {
+		return err
+	}
+	image, err := getImage(imageID, store)
```
Comment (mheon) [inline cmd/podman/rmi.go]: We have the image already, can't we pass this in?

File: cmd/podman/rmi.go [hunk 3]
```diff
@@ -77,3 +98,217 @@ func rmiCmd(c *cli.Context) error {
...
+	if err != nil || image == nil {
+		errors.Wrapf(err, "could not get image %q", imageID)
+	}
+
+	ctrIDs, err := runningContainers(image, store)
+	if err != nil {
+		return errors.Wrapf(err, "error getting running containers for image %q", imageID)
+	}
+	if len(ctrIDs) > 0 && len(image.Names) <= 1 {
```
Comment (mheon) [inline cmd/podman/rmi.go]: We don't need this logic, the standard logic in our image library for removing images should work fine

**Comment by mheon** [comment]
I think we really just need RunningContainers and RemoveContainers here. We can run RunningContainers unconditionally before we try and remove the image but after we remove Podman containers. If it returns that containers are present and --force is present, we use the RemoveContainers code to remove all non-Podman containers associated with the image. If it returns that containers are present and force is not set, we warn that non-Podman containers are using the image, and you should use --force if you want it gone. Then we continue with our removal logic. How does this sound?

**Comment by baude** [comment]
I agree with the proposed approach though part of me would prefer it be an import of buildah code ... nevertheless, I think that approach is spot on.

**Comment by TomSweeneyRedHat** [comment]
@baude @mheon thanks for the looksee and suggestions.  I'll fine tune.

File: cmd/podman/rmi.go [hunk 4]
```diff
@@ -86,3 +118,51 @@ func rmiCmd(c *cli.Context) error {
 	}
 	return lastError
 }
+
```
Comment (TomSweeneyRedHat) [inline cmd/podman/rmi.go]: I was waffling about keeping these following three functions in here or moving them into libpod/runtime_img.go

Comment (mheon) [reply cmd/podman/rmi.go]: I'd go ahead and move them. I am trying to keep the store entirely inside libpod, so directly opening a store and deleting outside of libpod is really undesirable. If we can make these part of runtime.RemoveImage, instead of the code here calling it, I think that'd be fine, and we wouldn't have to leak the store outside of libpod.

File: libpod/runtime_img.go [hunk 5]
```diff
@@ -124,7 +126,74 @@ func (r *Runtime) RemoveImage(image *image.Image, force bool) (string, error) {
 		return "", fmt.Errorf("unable to delete %s (must force) - image is referred to in multiple tags", image.ID())
 	}
 
-	return image.ID(), image.Remove(force)
+	err = image.Remove(force)
+	if err != nil && errors.Cause(err) == storage.ErrImageUsedByContainer {
```
Comment (mheon) [inline libpod/runtime_img.go]: I don't think we need this. We can just get storageContainers for the image and return an error if containers exist or remove them if --force is set. We don't need to do it in an error handling block, that makes things a lot harder to follow.

File: libpod/image/image.go [hunk 6]
```diff
@@ -78,6 +78,11 @@ func NewImageRuntimeFromOptions(options storage.StoreOptions) (*Runtime, error)
 	}, nil
 }
 
+// GetStore returns the store from storage
+func (i *Image) GetStore() storage.Store {
```
Comment (mheon) [inline libpod/image/image.go]: We should not need this. I really do not want to return a Store anywhere the user can get one.

File: libpod/runtime_img.go [hunk 7]
```diff
@@ -124,7 +126,74 @@ func (r *Runtime) RemoveImage(image *image.Image, force bool) (string, error) {
...
+		if errors.Cause(err) == storage.ErrImageUsedByContainer {
+			storageImage, err2 := r.getStorageImage(image.ID())
+			if err2 == nil {
+				errStorage := rmStorageContainers(force, storageImage, image.GetStore())
+				if errStorage == nil {
+					err = image.Remove(force)
+				} else {
+					fmt.Printf("A container created with Buildah may be associated with this image.\n")
```
Comment (mheon) [inline libpod/runtime_img.go]: No printing in libpod - return this as an error

File: cmd/podman/rmi.go [hunk 8]
```diff
@@ -53,12 +53,24 @@ func rmiCmd(c *cli.Context) error {
...
+	var imagesToCheck []*image.Image
 	var imagesToDelete []*image.Image
 	if removeAll {
-		imagesToDelete, err = runtime.ImageRuntime().GetImages()
+		imagesToCheck, err = runtime.ImageRuntime().GetImages()
 		if err != nil {
 			return errors.Wrapf(err, "unable to query local images")
 		}
+		// Look up the image by it's repository rather than its ID.
+		// We need this later if we need to remove for Storage (Buildah).
```
Comment (mheon) [inline cmd/podman/rmi.go]: Why is this necessary? Can we not find the image by its ID?

Comment (TomSweeneyRedHat) [reply cmd/podman/rmi.go]: I did this for the case when a image was tagged.  If I do:
```
podman pull alpine
podman tag alpine tg
podman images
REPOSITORY                 TAG      IMAGE ID       CREATED        SIZE
tg                         latest   3fd9065eaf02   2 months ago   4.41MB
docker.io/library/alpine   latest   3fd9065eaf02   2 months ago   4.41MB
```
Both "images" had the same ID, the only way I could differentiate them was to go by their repo.

File: libpod/runtime_img.go (hunk 7 above)
Comment (TomSweeneyRedHat) [reply libpod/runtime_img.go]: will do....

File: libpod/runtime_img.go (hunk 5 above)
Comment (TomSweeneyRedHat) [reply libpod/runtime_img.go]: I'm not sure exactly what you're suggesting here.  After line 129 are you saying I should just try getting the storage containers and then error if the image has containers and no force, other wise remove?  When we get to this point we've already tried removing the image and that failed due to it being a storage image and a container was associated.  I might be able to move the storage image handling that I've below into image.Remove() if that would be preferable.  I'll play with that and hit me up on IRC in the morn if I'm off base.

File: libpod/image/image.go (hunk 6 above)
Comment (TomSweeneyRedHat) [reply libpod/image/image.go]: If you or @baude have a suggestion around it, I'm happy to do it.  I tried using image.imageruntime.store in the call in runtime_img.go below  but got:

```
 github.com/projectatomic/libpod/libpod
_output/src/github.com/projectatomic/libpod/libpod/runtime_img.go:135:65: image.imageruntime undefined (cannot refer to unexported field or method imageruntime)
```

Thought exposing the function would be better.  However if I move that code to handle the storage containers into image.go as talked about below, that might solve that problem.

File: libpod/runtime_img.go (hunk 5 above)
Comment (mheon) [reply libpod/runtime_img.go]: Sorry, should have been more clear - I mean before 129, we can do the check if the container is mounted before the actual remove happens, thus removing the need to check errors.Cause() and perform complicated handling. This might make our fastpath a bit slower (we do a check unconditionally instead of conditionally), but will improve code flow and readability. We can always go back and make things faster later if need be.

File: libpod/image/image.go (hunk 6 above)
Comment (mheon) [reply libpod/image/image.go]: The image runtime's store is guaranteed to be the same as the normal runtime's store - you should just be able to runtime.store instead

File: cmd/podman/rmi.go [hunk 9]
```diff
@@ -53,12 +53,24 @@ func rmiCmd(c *cli.Context) error {
...
+		for _, img := range imagesToCheck {
+			for repo := range image.ReposToMap(img.Names()) {
```
Comment (baude) [inline cmd/podman/rmi.go]: @TomSweeneyRedHat im confused by this.  I'll catch you on IRC for follow-up

File: libpod/image/image.go (hunk 6 above)
Comment (baude) [reply libpod/image/image.go]: Agreed, i went to great pains to avoid this.

File: libpod/runtime_img.go [hunk 10]
```diff
@@ -124,7 +126,74 @@ func (r *Runtime) RemoveImage(image *image.Image, force bool) (string, error) {
...
+			} else {
+				fmt.Printf("Unable to find image %s\n", image.ID())
+			}
+		}
+	}
+	return image.ID(), err
+}
+
+// Remove containers that are in storage rather than Podman.
+func rmStorageContainers(force bool, image *storage.Image, store storage.Store) error {
```
Comment (baude) [inline libpod/runtime_img.go]: I would prefer to see this and subsequent functions take a *image.Image ... then you have the store and storage.Image already.  I say that assuming you at one point already had an *image.Image

File: libpod/runtime_img.go [hunk 11]
```diff
@@ -154,3 +224,25 @@ func GetInsecureRegistries() ([]string, error) {
 	}
 	return registries, nil
 }
+
+// GetStorageImage retrieves an image matching the given name or hash from
+// system storage.
+// If no matching image can be found, an error is returned.
```
Comment (baude) [inline libpod/runtime_img.go]: Ill hit you up on irc for clarification on this func as well.

File: libpod/image/image.go (hunk 6 above)
Comment (TomSweeneyRedHat) [reply libpod/image/image.go]: ack, I was over complicating it, I can just do r.store in runtime_img.go and it gets me what I need.  Thought I'd to pull it from each image and that wasn't resolving.

File: cmd/podman/rmi.go (hunk 9 above)
Comment (TomSweeneyRedHat) [reply cmd/podman/rmi.go]: @baude and I worked through this before scrum.  The issue I was running into was deleting container from an image created by Buildah when the image had a tag added to it.  The code as I had it here worked around that, however @baude pointed to a tweak of a if statement down lower that let me remove this code and still get the containers removed.  This has all been removed.

File: libpod/runtime_img.go (hunk 11 above)
Comment (TomSweeneyRedHat) [reply libpod/runtime_img.go]: We only talked briefly about this, but I've removed this as you indicated elsewhere I can get the info here from the runtime (r).

File: libpod/runtime_img.go (hunk 5 above)
Comment (TomSweeneyRedHat) [reply libpod/runtime_img.go]: This chunk has been reworked and greatly reduced.  I still have the error check, but I think it makes more sense now.  Cand/will adjust after the next review if you think appropriate.

File: libpod/runtime_img.go (hunk 10 above)
Comment (TomSweeneyRedHat) [reply libpod/runtime_img.go]: reworking.  I've passed image.Image to the rmStorageContainers and just the id to one other function rather than the whole image.

File: cmd/podman/rmi.go [hunk 12]
```diff
@@ -70,12 +71,16 @@ func rmiCmd(c *cli.Context) error {
...
 	}
+
 	if len(imagesToDelete) == 0 {
 		return errors.Errorf("no valid images to delete")
 	}
 	for _, img := range imagesToDelete {
 		msg, err := runtime.RemoveImage(img, c.Bool("force"))
 		if err != nil {
+			if errors.Cause(err) == storage.ErrImageUsedByContainer {
+				fmt.Printf("A container associated with Buildah may be associated with this image.\n")
```
Comment (mheon) [inline cmd/podman/rmi.go]: Shouldn't just call out buildah here - could also be CRI-O.

Comment (mheon) [reply cmd/podman/rmi.go]: Ideally, should mention containers/storage specifically, could be a program using c/storage we didn't write

File: libpod/runtime_img.go [hunk 13]
```diff
@@ -123,8 +124,61 @@ func (r *Runtime) RemoveImage(image *image.Image, force bool) (string, error) {
 		// reponames and no force is applied, we error out.
 		return "", fmt.Errorf("unable to delete %s (must force) - image is referred to in multiple tags", image.ID())
 	}
+	err = image.Remove(force)
+	if err != nil && errors.Cause(err) == storage.ErrImageUsedByContainer {
+		errStorage := r.rmStorageContainers(force, image)
+		if errStorage == nil {
+			err = image.Remove(force)
```
Comment (mheon) [inline libpod/runtime_img.go]: Are we retrying on the assumption that the containers using the image were removed? If so, can we get a comment to that effect?

Comment (mheon) [reply libpod/runtime_img.go]: Also, instead of setting error, it's probably better to just `return` here and below in the else

File: cmd/podman/rmi.go (hunk 12 above)
Comment (baude) [reply cmd/podman/rmi.go]: Couple of things here.  First, if the err is no ErrImageUserByContainer, isn't the error lost?  And second, should your error message also contain the contents of err?

File: libpod/runtime_img.go (hunk 13 above)
Comment (baude) [reply libpod/runtime_img.go]: is there any chance that err might not be ErrImageUsedByContainer ?

File: libpod/runtime_img.go [hunk 14]
```diff
@@ -123,8 +124,61 @@ func (r *Runtime) RemoveImage(image *image.Image, force bool) (string, error) {
...
+
+// Remove containers that are in storage rather than Podman.
+func (r *Runtime) rmStorageContainers(force bool, image *image.Image) error {
+
+	ctrIDs, err := storageContainers(image.ID(), r.store)
+	if err != nil {
+		return errors.Wrapf(err, "error getting containers for image %q", image.ID())
+	}
+
+	if len(ctrIDs) > 0 {
```
Comment (baude) [inline libpod/runtime_img.go]: && force?

File: libpod/runtime_img.go [hunk 15]
```diff
@@ -123,8 +124,61 @@ func (r *Runtime) RemoveImage(image *image.Image, force bool) (string, error) {
...
+		if force {
+			err = removeStorageContainers(ctrIDs, r.store)
```
Comment (baude) [inline libpod/runtime_img.go]: if err := remove.... ; err != nil

**Comment by mheon** [comment]
@TomSweeneyRedHat can you rebase on latest master? The test failures should be fixed

File: cmd/podman/rmi.go (hunk 12 above)
Comment (TomSweeneyRedHat) [reply cmd/podman/rmi.go]: I've changed the error message, bumped it a bit from what @mheon suggested.  I don't think the error is lost here, we save it as we do currently in lastError.  This only prints out extra verbiage when we have a container tied to the image.  The error that prints I didn't think added a lot here, and the error message is currently pretty long so I didn't include it.  If you've a strong opinion on adding it I can.

File: libpod/runtime_img.go (hunk 13 above)
Comment (TomSweeneyRedHat) [reply libpod/runtime_img.go]: I've added a comment.

I think we want to capture the err from this second remove and return it if there's an error.  But if you really think it's better, I can just let the original err be returned.  As far as returning the errStorage error I was on the fence about returning that or not, but thought it would give a better clue if there was an error.  I'll keep them both for now, holler in he morn if you'd like me to remove them.

I think there's a very small chance that the err would be something other then ErrImageUsedByContainer.   But if it is, we'll just get the same error from the second image remove and will continue on down the error path.  But for the vast majority, it should be the image in use error.

File: libpod/runtime_img.go (hunk 14 above)
Comment (TomSweeneyRedHat) [reply libpod/runtime_img.go]: good call, done.

File: libpod/runtime_img.go (hunk 15 above)
Comment (TomSweeneyRedHat) [reply libpod/runtime_img.go]: yep, found a second one too, thx.

**Comment by TomSweeneyRedHat** [comment]
#564 should hopefully fix up the test errors here.

**Comment by rhatdan** [comment]
@TomSweeneyRedHat Rebase and try.

File: cmd/podman/rmi.go [hunk 16]
```diff
@@ -76,6 +77,9 @@ func rmiCmd(c *cli.Context) error {
 	for _, img := range imagesToDelete {
 		msg, err := runtime.RemoveImage(img, c.Bool("force"))
 		if err != nil {
+			if errors.Cause(err) == storage.ErrImageUsedByContainer {
+				fmt.Printf("A container associated with containers/storage, i.e. via Buildah or CRI-O,  may be associated with this image.\n")
```
[the remark this answers is not in the document]
Comment (TomSweeneyRedHat) [reply cmd/podman/rmi.go]: I hear you, but my main concern that if a customer got an error that there's a "container associated with containers/storage" would they know enough to check Buildah/CRI-O?  I've tried to open it up a bit by adding etc.

File: libpod/runtime_img.go [hunk 17]
```diff
@@ -133,6 +185,7 @@ func GetRegistries() ([]string, error) {
 	envOverride := os.Getenv("REGISTRIES_CONFIG_PATH")
 	if len(envOverride) > 0 {
 		registryConfigPath = envOverride
+
```
[the remark this answers is not in the document]
Comment (TomSweeneyRedHat) [reply libpod/runtime_img.go]: Done, don't know how I managed that one.

**Comment by rhatdan** [comment]
@mheon @baude @umohnani8 PTAL
LGTM

File: libpod/runtime_img.go [hunk 18]
```diff
@@ -123,8 +124,59 @@ func (r *Runtime) RemoveImage(image *image.Image, force bool) (string, error) {
...
+// Remove containers that are in storage rather than Podman.
+func (r *Runtime) rmStorageContainers(force bool, image *image.Image) error {
 
-	return image.ID(), image.Remove(force)
+	ctrIDs, err := storageContainers(image.ID(), r.store)
+	if err != nil {
+		return errors.Wrapf(err, "error getting containers for image %q", image.ID())
+	}
+
+	if len(ctrIDs) > 0 && force {
```
Comment (mheon) [inline libpod/runtime_img.go]: Shouldn't we throw ErrImageUsedByContainer here if `len(ctrIDs) > 0 && !force`?

File: libpod/runtime_img.go [hunk 19]
```diff
@@ -123,8 +124,59 @@ func (r *Runtime) RemoveImage(image *image.Image, force bool) (string, error) {
...
+		if err = removeStorageContainers(ctrIDs, r.store); err != nil {
+			return errors.Wrapf(err, "error removing containers %v for image %q", ctrIDs, image.ID())
+		}
+	}
+	return nil
+}
+
+// Returns a list of storage containers associated with the given ImageReference
+func storageContainers(imageID string, store storage.Store) ([]string, error) {
+
```
Comment (mheon) [inline libpod/runtime_img.go]: Extra whitespace

Comment (TomSweeneyRedHat) [reply libpod/runtime_img.go]: Fixed, found another rogue one and zapped that too.

**Comment by umohnani8** [comment]
LGTM

File: libpod/runtime_img.go (hunk 18 above)
Comment (TomSweeneyRedHat) [reply libpod/runtime_img.go]: I was relying on the second image.Remove(force) to provide that error at line 132.  However it probably does make more sense to throw the error here and not invoke that image.Remove() again as we know it will fail in this instance.  Good thought.

**Comment by mheon** [comment]
Changes LGTM
@rh-atomic-bot r+

**Comment by TomSweeneyRedHat** [comment]
@baude  @rhatdan @mheon I've just posted a new patch.  I fixed up @mheon's comments and while testing noted an error message in rmi.go that would be better served with the image.ID() being printed.

**Comment by TomSweeneyRedHat** [comment]
Quick test output:
```
# buildah from alpine
Getting image source signatures
Copying blob sha256:ff3a5c916c92643ff77519ffa742d3ec61b7f591b6b7504599d95a4a41134e28
 1.97 MiB / 1.97 MiB [=====================================================] 21s
Copying config sha256:3fd9065eaf02feaf94d68376da52541925650b81698c53c6824d92ff63f98353
 1.48 KiB / 1.48 KiB [======================================================] 0s
Writing manifest to image destination
Storing signatures
alpine-working-container

# podman pull busybox
Trying to pull docker.io/busybox:latest...Getting image source signatures
Copying blob sha256:f70adabe43c0cccffbae8785406d490e26855b8748fc982d14bc2b20c778b929
 706.22 KB / 706.22 KB [====================================================] 0s
Copying config sha256:8ac48589692a53a9b8c2d1ceaa6b402665aa7fe667ba51ccc03002300856d8c7
 1.46 KB / 1.46 KB [========================================================] 0s
Writing manifest to image destination
Storing signatures
8ac48589692a53a9b8c2d1ceaa6b402665aa7fe667ba51ccc03002300856d8c7

# podman tag alpine tg

# podman ps -a

# podman images
REPOSITORY                  TAG      IMAGE ID       CREATED        SIZE
tg                          latest   3fd9065eaf02   2 months ago   4.41MB
docker.io/library/alpine    latest   3fd9065eaf02   2 months ago   4.41MB
docker.io/library/busybox   latest   8ac48589692a   5 hours ago    1.36MB

# podman rmi -a
8ac48589692a53a9b8c2d1ceaa6b402665aa7fe667ba51ccc03002300856d8c7
unable to delete 3fd9065eaf02feaf94d68376da52541925650b81698c53c6824d92ff63f98353 (must force) - image is referred to in multiple tags

# podman images
REPOSITORY                 TAG      IMAGE ID       CREATED        SIZE
docker.io/library/alpine   latest   3fd9065eaf02   2 months ago   4.41MB
tg                         latest   3fd9065eaf02   2 months ago   4.41MB

# podman rmi tg
Untagged: tg:latest

# podman rmi -a
A container associated with containers/storage, i.e. via Buildah, CRI-O, etc., may be associated with this image: 3fd9065eaf02
image is in use by a container

# podman rmi -a -f
3fd9065eaf02feaf94d68376da52541925650b81698c53c6824d92ff63f98353

```

**Comment by TomSweeneyRedHat** [comment]
Addresses #174
